rpmio/rpmkeyring.c

```diff
@@ -153,37 +155,83 @@ rpmPubkey rpmPubkeyNew(const uint8_t *pkt, size_t pktlen)
     return key;
 }
 
-static rpmPubkey rpmPubkeyNewSubkey(pgpDigParams pgpkey)
+static rpmPubkey rpmPubkeyNewSubkey(rpmPubkey primarykey, pgpDigParams pgpkey)
 {
     rpmPubkey key = new rpmPubkey_s {};
-    /* Packets with all subkeys already stored in main key */
+    /* Packets with all subkeys already stored in primary key */
+    key->primarykey = primarykey;
     key->pgpkey = pgpkey;
     key->keyid = key2str(pgpDigParamsSignID(pgpkey));
     key->nrefs = 1;
     return key;
 }
 
-rpmPubkey *rpmGetSubkeys(rpmPubkey mainkey, int *count)
+rpmPubkey *rpmGetSubkeys(rpmPubkey primarykey, int *count)
 {
     rpmPubkey *subkeys = NULL;
     pgpDigParams *pgpsubkeys = NULL;
     int pgpsubkeysCount = 0;
     int i;
 
-    if (mainkey && !pgpPrtParamsSubkeys(mainkey->pkt.data(), mainkey->pkt.size(),
-			mainkey->pgpkey, &pgpsubkeys, &pgpsubkeysCount)) {
+    wrlock lock(primarykey->mutex);
+
+    if (primarykey && !pgpPrtParamsSubkeys(primarykey->pkt.data(), primarykey->pkt.size(),
```

Comment:
Uff. There's a booby-trap in there that I'd totally missed: we're testing whether primarykey is NULL, but we already accessed primarykey->mutex in the above so the compiler might go optimize-happy doing who knows what. Sorry for missing this earlier, the whole rpmGetSubkeys() thing got tacked on the side of this PR when it looked like just a trivial thing but of course it wasn't *so* trivial afterall.

Reply:
Ooofff.

I wonder if that function should be there in the first place. Shouldn't we just add the subkeys automatically when adding the primary keys?

Reply:
Indeed, this function doesn't make a whole lot of sense to begin with, rpmKeyringAddKey() should just do this without requiring all users go through this dance.

Which makes the scope ever that bigger - I would say just drop the Subkeys locking part from this PR - it was always broken and clearly could use some sanity instead of locking.

Reply:
I'll file a ticket for that so we don't forget.

Reply:
That became two: #3350 and #3351

Reply:
To me it looks like moving the `if(primarykey` up should do the trick.

Reply:
It should yeah, so if you want to do that here, fine, but this looks more like the whole approach of subkeys needs to be looked at, and I don't want us to waste more time fiddling, reviewing and fiddling and reviewing if that thing might be gone in the "next" PR.

Reply:
Ok, reverted to the previous version (only renaming "mainkey" to primarykey")